java-checks/src/main/java/org/sonar/java/checks/naming/BadConstantNameCheck.java

```diff
@@ -89,19 +114,50 @@ private void checkName(VariableTree variableTree) {
     }
   }
 
-  private static boolean isStaticFinal(VariableTree variableTree) {
-    boolean isStatic = false;
-    boolean isFinal = false;
-    for (ModifierKeywordTree modifierKeywordTree : variableTree.modifiers().modifiers()) {
-      Modifier modifier = modifierKeywordTree.modifier();
-      if (modifier == Modifier.STATIC) {
-        isStatic = true;
-      }
-      if (modifier == Modifier.FINAL) {
-        isFinal = true;
+  private static boolean isStaticFinal(Symbol variable) {
+    return (variable.isStatic() && variable.isFinal())
+      || (variable.owner().isMethodSymbol() && variable.isFinal());
+  }
+
+  private static class VariableFromMethodVisitor extends BaseTreeVisitor {
+
+    private final MethodTree method;
+    private final BadConstantNameCheck check;
```

Comment:
you could make the inner class not static instead of passing around the check, that's the purpose of inner class.